nowcasting_dataset/data_sources/optical_flow/optical_flow_data_source.py

```diff
@@ -0,0 +1,350 @@
+        return OpticalFlow
+
+    def _put_predictions_into_data_array(
+        self,
+        satellite_data: xr.DataArray,
+        predictions: np.ndarray,
+    ) -> xr.DataArray:
+        """
+        Puts optical flow predictions into an xr.DataArray.
+
+        Args:
+            satellite_data: Satellite data
+            predictions: Predictions from the optical flow
+
+        Returns:
+            The Xarray DataArray with the optical flow predictions
+        """
+        # Generate a pd.DatetimeIndex for the optical flow predictions.
+        t0_datetime_utc = satellite_data.isel(time=-1)["time"].values
+        t1_datetime_utc = t0_datetime_utc + self.sample_period_duration
+        datetime_index_of_predictions = pd.date_range(
+            t1_datetime_utc, periods=self.forecast_length, freq=self.sample_period_duration
+        )
+
+        # Select the center crop.
+        satellite_data_cropped = satellite_data.isel(time=0, channels=0)
+        satellite_data_cropped = crop_center(satellite_data_cropped, self.output_image_size_pixels)
+
+        # Put into DataArray:
+        return xr.DataArray(
+            data=predictions,
+            coords=(
+                ("time", datetime_index_of_predictions),
+                ("x", satellite_data_cropped.coords["x"].values),
+                ("y", satellite_data_cropped.coords["y"].values),
+                ("channels", satellite_data.coords["channels"].values),
+            ),
+            name="data",
+        )
+
+    def _compute_and_return_optical_flow(self, satellite_data: xr.DataArray) -> xr.DataArray:
+        """
+        Compute and return optical flow predictions for the example
+
+        Args:
+            satellite_data: Satellite DataArray of historical satellite images, up to and include t0
+
+        Returns:
+            DataArray with the optical flow predictions from t1 to the forecast horizon.
+        """
+        n_channels = satellite_data.sizes["channels"]
+
+        # Sanity check
+        assert (
+            len(satellite_data.coords["time"]) == self.history_length + 1
+        ), f"{len(satellite_data.coords['time'])=} != {self.history_length+1=}"
+        assert n_channels == len(self.channels), f"{n_channels=} != {len(self.channels)=}"
+
+        # Pre-allocate an array, into which our optical flow prediction will be placed.
+        prediction_block = np.full(
+            shape=(
+                self.forecast_length,
+                self.output_image_size_pixels,
+                self.output_image_size_pixels,
+                n_channels,
+            ),
+            fill_value=-1,
+            dtype=np.int16,
+        )
+
+        # Compute flow fields and optical flow predictions separately for each satellite channel
+        # because the different channels represent different physical phenomena and so,
+        # in principle, could move in different directions (e.g. water vapour vs high clouds).
+        for channel_i in range(n_channels):
```

Comment:
If this is taking a long time? Do we need to speed up these loops somehow?

Reply:
It's fast enough for now, I think.  It takes a few hours to create 4,000 batches.  If needs be, we could speed this up by running on multiple CPU cores.  (indeed, my original code for optical flow used Python 3.8's fancy shared-memory-across-processes stuff, and it was really fast.. but also a bit fiddly to debug!)

Reply:
Cool - lets keep it how it is then